include/pybind11/pybind11.h

```diff
@@ -2153,7 +2158,14 @@ class gil_scoped_release {
     ~gil_scoped_release() {
         if (!tstate)
             return;
+#if PY_VERSION_HEX >= 0x03070000
+        // PyEval_RestoreThread() should not be called if runtime is finilizing
+        // See https://docs.python.org/3/c-api/init.html#c.PyEval_RestoreThread
+        if (!_Py_IsFinalizing())
+            PyEval_RestoreThread(tstate);
+#else
```

Comment:
Minor: typo: finilizing

More significantly: I think the code could be made more readable and maintainable by centralizing the `#if #else #endif` in a, e.g., `detail::finalization_guard()` inline helper function, so that there is a formal link to the comment from both places, and less preprocessor clutter.

Reply:
Sure that that's not just going to add more confusion, to have another indirection? As long as it's just these two (very related) spots, to me, this seem reasonably fine? (We currently have these kinds of things inline in other places as well, and this should make it easy to remove once we drop <= 3.6?)

Reply:
The practical version of DRY/SPOT that I've seen states that exactly one duplication is allowed, if it's clearly more complex to combine it. Trying to take DRY/SPOT to the extreme of truly one point of truth for all code can impose an unmanageable level of complexity, so allowing it to be relaxed by one at programmer discretion balances the ideal with practicality. (and if there's more than one duplication, the extra complexity is always worth it).

In short, I'm fine either way.

Reply:
Thanks Henry! I hadn't heard of DRY/SPOT before, good to know there is something to refer to.

In this particular case I'm not so much concerned about the code duplication, but mostly about the comment only appearing in one place. For someone arriving at the second place from some completely different context, they will not even know the comment exists in the other place, and may miss it. But if they get curious about `detail::finalization_guard()`, they are guaranteed to find the comment.

Reply:
The comment is true, though! I completely agree there. There should at least be a "see a few lines up/down, at ...".

Apart from that, I don't feel too strongly, so up to @malfet AFAIC, but I just though the `#if SOME_PYTHON_VERSION` pattern exists inline in lots of places of code.
And it's kind of hard to capture what it's doing (i.e., `finalization_guard` doesn't really describe what's happening) ?

Reply:
Thank you for the feedback. Fixed typo and moved and extended runtime finalization check to `pybind11/detail/internals.h`